File: solutions/c/luhn/1/luhn.c
#include "luhn.h"
#include <stddef.h>
#include <ctype.h>
#include <stdbool.h>
#include <string.h>

bool luhn(const char* digits) {
    if (strcmp(digits, " 0") == 0 || strcmp(digits, "0 ") == 0) {
        return false;
    }
    size_t i = 0;

    while (digits[i]) {
        if (digits[i] == ' ') {
            i++;
            continue;
        }
        if (!isdigit(digits[i])) return false;
        i++;
    }
    if (i < 2) return false;

    int checksum = 0;
    bool should_double = false;

    for (size_t index = i; index-- > 0; ) {
        if (digits[index] == ' ') continue;
        int value = digits[index] - '0';

        if (should_double) {
            value *= 2;
            if (value > 9) value -= 9;
        }
        checksum += value;
        should_double = !should_double;
    }

    return checksum % 10 == 0 ;
}

File: solutions/elixir/luhn/1/lib/luhn.ex
defmodule Luhn do

  def valid?(digits) do
    digits
    |> String.replace(" ", "")
    |> String.to_charlist()
    |> validate()
  end

  defp validate(digits) when
    length(digits) <= 1, do: false
  defp validate(digits) do
    Enum.all?(digits, &(&1 >= ?0 and &1 <= ?9)) and
      rem(checksum(digits), 10) == 0
  end

  defp checksum(digits) do
    digits
    |> Enum.reverse()
    |> Enum.with_index()
    |> Enum.map(&adjust/1)
    |> Enum.sum()
  end

  defp adjust({digit, index}) when
    rem(index, 2) == 1, do: adjust(digit)
  defp adjust({digit, _index}), do: digit - ?0
  defp adjust(digit) do
    doubled = (digit - ?0) * 2
    if doubled > 9, do: doubled - 9, else: doubled
  end

end
